src/tests/t_kdb.py

```diff
@@ -336,6 +336,31 @@ def ldap_add(dn, objectclass, attrs=[]):
 realm.kinit('canon', password('canon'))
 realm.kinit('alias', password('canon'), ['-C'])
 
+# Test password history
+def test_pwhist(nhist):
+    output('Testing password history of size ' + `nhist`)
+    princ = 'pwhistprinc' + `nhist`
+    pol = 'pwhistpol'+`nhist`
+    realm.run([kadminl, 'addpol', '-history', `nhist`, pol])
+    realm.run([kadminl, 'addprinc', '-policy', pol, '-nokey', princ])
+    # Test nhist number of passwords and make sure that all previous passwords
+    # fail when trying to be changed to them.
+    cpw = lambda x,y: realm.run([kadminl, 'cpw', '-pw', `x`, princ],
+                                expected_code=y)
+
+    # Creates a combination of password attempts, with (x, 0) and ([0..x], 1)
+    cmb = lambda x: izip(chain([x], islice(count(), x+1)),
+                             chain([0], repeat(1)))
+    # Tries passwords, with one password change that should succeed followed by
+    # all previous passwords that should fail
+    # (0, 0) (0, 1) (1, 0) (0, 1), (1, 1)...
+    list(starmap(cpw, chain.from_iterable(imap(cmb, islice(count(), nhist)))))
+
+    # Test that keys are rotated out correctly
+    cpw(nhist, 0)
+    cpw(0, 0)
+map(test_pwhist, imap(pow, repeat(2), range(5)))
```

Comment:
Here is what I would consider a more reasonable version of the same tests, in the same number of lines of Python code:

    # Test password history.                                                    
    def test_pwhist(nhist):
        def cpw(n):
            realm.run([kadminl, 'cpw', '-pw', str(n), princ])
        def cpw_fail(n):
            realm.run([kadminl, 'cpw', '-pw', str(n), princ], expected_code=1)
        output('*** Testing password history of size %d\n' % nhist)
        princ = 'pwhistprinc' + str(nhist)
        pol = 'pwhistpol' + str(nhist)
        realm.run([kadminl, 'addpol', '-history', str(nhist), pol])
        realm.run([kadminl, 'addprinc', '-policy', pol, '-nokey', princ])
        for i in range(nhist):
            # Set a password, then check that all previous passwords fail.      
            cpw(i)
            for j in range(i + 1):
                cpw_fail(i)
        # Set one more new password, and make sure the oldest key is            
        # rotated out.                                                          
        cpw(nhist)
        # I would add: cpw_fail(1)
        cpw(0)

    for n in (1, 2, 4, 8, 16):
        test_pwhist(n)

I can understand being concerned by the complexity of add_to_history(); that function is grossly overcomplicated because it uses a circular queue for no reason.  (The old keys are marshalled and unmarshalled on each password change, which is O(n) in the number of old keys, so trying to keep the insertion cost to O(1) is pointless.)  I would be happy to see it rewritten to be simpler.    But for its current incarnation, running a big quadratic test doesn't hit any more of its edge cases than a small quadratic test.  To cover that function properly you would need to do things like change the password policy after loading up the principal with some old keys.